File: backend/models/note.go
package models

import "time"

type Note struct {
	ID              string    `db:"id"              json:"id"`
	UserID          string    `db:"user_id"         json:"user_id"`
	Title           string    `db:"title"           json:"title"`
	Body            string    `db:"body"            json:"-"` // never directly in JSON
	IsSecret        bool      `db:"is_secret"       json:"is_secret"`
	BodyNonce       *string   `db:"body_nonce"      json:"-"`
	IsPinned        bool      `db:"is_pinned"       json:"is_pinned"`
	Color           string    `db:"color"           json:"color"`
	CreatedAt       time.Time `db:"created_at"      json:"created_at"`
	UpdatedAt       time.Time `db:"updated_at"      json:"updated_at"`
	AttachmentCount int       `db:"attachment_count" json:"attachment_count"`
}

// NoteResponse is returned to clients. Body is a pointer: nil when secret+locked.
type NoteResponse struct {
	ID              string    `json:"id"`
	UserID          string    `json:"user_id"`
	Title           string    `json:"title"`
	Body            *string   `json:"body"` // nil when secret and not unlocked
	IsSecret        bool      `json:"is_secret"`
	IsLocked        bool      `json:"is_locked"` // true when secret and not yet unlocked in this request
	IsPinned        bool      `json:"is_pinned"`
	Color           string    `json:"color"`
	CreatedAt       time.Time `json:"created_at"`
	UpdatedAt       time.Time `json:"updated_at"`
	AttachmentCount int       `json:"attachment_count"`
}

func (n *Note) ToResponse(decryptedBody *string) NoteResponse {
	resp := NoteResponse{
		ID:              n.ID,
		UserID:          n.UserID,
		Title:           n.Title,
		IsSecret:        n.IsSecret,
		IsLocked:        n.IsSecret && decryptedBody == nil,
		IsPinned:        n.IsPinned,
		Color:           n.Color,
		CreatedAt:       n.CreatedAt,
		UpdatedAt:       n.UpdatedAt,
		AttachmentCount: n.AttachmentCount,
	}
	if !n.IsSecret {
		resp.Body = &n.Body
	} else {
		resp.Body = decryptedBody // nil if locked
	}
	return resp
}

File: mobile/lib/models/note.dart
import 'tag.dart';

class Note {
  final String id;
  final String userId;
  final String title;
  final String? body; // null when secret + locked
  final bool isSecret;
  final bool isLocked;
  final bool isPinned;
  final String? color;
  final List<Tag> tags;
  final String createdAt;
  final String updatedAt;
  final int? attachmentCount;

  const Note({
    required this.id,
    required this.userId,
    required this.title,
    this.body,
    required this.isSecret,
    required this.isLocked,
    required this.isPinned,
    this.color,
    required this.tags,
    required this.createdAt,
    required this.updatedAt,
    this.attachmentCount,
  });

  factory Note.fromJson(Map<String, dynamic> j) => Note(
        id: j['id'] as String,
        userId: j['user_id'] as String? ?? '',
        title: j['title'] as String? ?? '',
        body: j['body'] as String?,
        isSecret: (j['is_secret'] as bool?) ?? false,
        isLocked: (j['is_locked'] as bool?) ?? false,
        isPinned: (j['is_pinned'] as bool?) ?? false,
        color: j['color'] as String?,
        tags: (j['tags'] as List<dynamic>?)
                ?.map((t) => Tag.fromJson(t as Map<String, dynamic>))
                .toList() ??
            [],
        createdAt: j['created_at'] as String? ?? '',
        updatedAt: j['updated_at'] as String? ?? '',
        attachmentCount: (j['attachment_count'] as num?)?.toInt(),
      );

  Note copyWith({
    String? title,
    String? body,
    bool clearBody = false,
    bool? isSecret,
    bool? isLocked,
    bool? isPinned,
    String? color,
    List<Tag>? tags,
    String? updatedAt,
    int? attachmentCount,
  }) =>
      Note(
        id: id,
        userId: userId,
        title: title ?? this.title,
        body: clearBody ? null : (body ?? this.body),
        isSecret: isSecret ?? this.isSecret,
        isLocked: isLocked ?? this.isLocked,
        isPinned: isPinned ?? this.isPinned,
        color: color ?? this.color,
        tags: tags ?? this.tags,
        createdAt: createdAt,
        updatedAt: updatedAt ?? this.updatedAt,
        attachmentCount: attachmentCount ?? this.attachmentCount,
      );
}
